### DISCUSSION TITLE
Failing to get started with a fresh installation...

### DISCUSSION THREAD

**Discussion by Canowyrms** [payloadcms/payload#3783] [Q&A]
### Describe the Bug

Hello. I'm trying to spin up Payload to play around with it. To the best of my knowledge, I'm following the installation instructions as they're written, but I'm encountering some issues... I did search here and Google before writing this, but I wasn't able to find anything related :\

Node: v18.18.2
NPM: 10.2.0
Yarn: 4.0.0-rc.53
(Above installed via Volta, if that matters.)
Ubuntu 22.04.03 LTS on WSL2 on Win10.
Docker Desktop: v4.24.2 (using WSL2 backend)

---

Not sure if this is an issue with Payload itself or with Yarn/whatever package manager it decides to use with `npx create-payload-app` (it's not clear to me which one is being used during this process). After creating the Payload app and `cd`ing in, I run `yarn dev`, like it says, and I get some error output that the app crashed.

<details>
<summary>Expand for terminal output</summary>

```sh
bk@wsl ~/code/sandbox
$ npx create-payload-app -n payload-cms-sandbox -t website

  Welcome to Payload. Let's create a project!

✔ Select a database › MongoDB
✔ Enter MongoDB connection string … mongodb://127.0.0.1/payload-cms-sandbox

  Creating project in /home/bk/code/sandbox/payload-cms-sandbox

✔ Dependencies installed
✔ .env file created
✔ Payload project successfully created

  ★ Launch Application:

    - cd ./payload-cms-sandbox
    - yarn dev or follow directions in README.md: file:///home/bk/code/sandbox/payload-cms-sandbox/README.md

  ★ Documentation:

    - Getting Started: https://payloadcms.com/docs/getting-started/what-is-payload
    - Configuration: https://payloadcms.com/docs/configuration/overview



bk@wsl ~/code/sandbox
$ cd payload-cms-sandbox/

bk@wsl ~/code/sandbox/payload-cms-sandbox
$ yarn dev
[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: ts
[nodemon] starting `ts-node src/server.ts -- -I`
unhandledRejection Error: mongoose-aggregate-paginate-v2 tried to access mongoose, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.

Required package: mongoose
Required by: mongoose-aggregate-paginate-v2@npm:1.0.6 (via /home/bk/.yarn/berry/cache/mongoose-aggregate-paginate-v2-npm-1.0.6-df5a3d46e2-10c0.zip/node_modules/mongoose-aggregate-paginate-v2/)

Require stack:
- /home/bk/.yarn/berry/cache/mongoose-aggregate-paginate-v2-npm-1.0.6-df5a3d46e2-10c0.zip/node_modules/mongoose-aggregate-paginate-v2/index.js
- /home/bk/code/sandbox/payload-cms-sandbox/.yarn/__virtual__/@payloadcms-db-mongodb-virtual-4fb9dbaf43/4/.yarn/berry/cache/@payloadcms-db-mongodb-npm-1.0.4-5a642e2349-10c0.zip/node_modules/@payloadcms/db-mongodb/dist/init.js
- /home/bk/code/sandbox/payload-cms-sandbox/.yarn/__virtual__/@payloadcms-db-mongodb-virtual-4fb9dbaf43/4/.yarn/berry/cache/@payloadcms-db-mongodb-npm-1.0.4-5a642e2349-10c0.zip/node_modules/@payloadcms/db-mongodb/dist/index.js
- /home/bk/code/sandbox/payload-cms-sandbox/src/payload/payload.config.ts
- /home/bk/.yarn/berry/cache/payload-npm-2.0.11-0230c0b9cd-10c0.zip/node_modules/payload/dist/config/load.js
- /home/bk/.yarn/berry/cache/payload-npm-2.0.11-0230c0b9cd-10c0.zip/node_modules/payload/dist/payload.js
- /home/bk/.yarn/berry/cache/payload-npm-2.0.11-0230c0b9cd-10c0.zip/node_modules/payload/dist/initHTTP.js
- /home/bk/.yarn/berry/cache/payload-npm-2.0.11-0230c0b9cd-10c0.zip/node_modules/payload/dist/index.js
- /home/bk/code/sandbox/payload-cms-sandbox/src/server.ts
    at Function.require$$0.Module._resolveFilename (/home/bk/code/sandbox/payload-cms-sandbox/.pnp.cjs:22531:13)
    at Function.call (/home/bk/.yarn/berry/cache/@cspotcode-source-map-support-npm-0.8.1-964f2de99d-10c0.zip/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at /home/bk/code/sandbox/payload-cms-sandbox/.yarn/__virtual__/next-virtual-afab771527/4/.yarn/berry/cache/next-npm-13.5.6-306005c39c-10c0.zip/node_modules/next/src/server/require-hook.ts:50:34
    at Function.Module._load (node:internal/modules/cjs/loader:922:27)
    at Function.require$$0.Module._load (/home/bk/code/sandbox/payload-cms-sandbox/.pnp.cjs:22422:31)
    at Module.call (node:internal/modules/cjs/loader:1143:19)
    at Module.mod.require (/home/bk/code/sandbox/payload-cms-sandbox/.yarn/__virtual__/next-virtual-afab771527/4/.yarn/berry/cache/next-npm-13.5.6-306005c39c-10c0.zip/node_modules/next/src/server/require-hook.ts:69:26)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (/home/bk/.yarn/berry/cache/mongoose-aggregate-paginate-v2-npm-1.0.6-df5a3d46e2-10c0.zip/node_modules/mongoose-aggregate-paginate-v2/index.js:1:18)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
[nodemon] app crashed - waiting for file changes before starting...
```
</details>

---

I also tried to get started by using the included `docker-compose.yml`... I had some problems there, too.

```sh
bk@wsl ~/code/sandbox/payload-cms-sandbox
$ docker-compose up
[truncated]
payload-cms-sandbox-payload-1  | [03:12:18] ERROR (payload): Error: cannot connect to MongoDB. Details: connect ECONNREFUSED 127.0.0.1:27017
```

Over in the browser, hitting `http://127.0.0.1:3000` returns an error:

> The connection to the server was reset while the page was loading.

---

What am I doing wrong? How am I managing to fail two different ways?

### To Reproduce

1. Run `npx create-payload-app -n payload-cms-sandbox -t website`
2. Run `cd payload-cms-sandbox`
3. Run `yarn dev`
4. App crashes

---

1. Run `npx create-payload-app -n payload-cms-sandbox -t website`
2. Run `cd payload-cms-sandbox`
3. Run `docker-compose up`
4. Watch terminal for error trying to connect to MongoDB.
5. Request `http://127.0.0.1:3000` in the browser, receive 'connection reset' error.

### Payload Version

Latest?

**Comment by JessRynkar** [comment]
Hi @Canowyrms - I can help you figure what is going on here! Thanks for providing all the details about your machine and software. 

The steps you followed are correct, but it appears MongoDB is not connecting for you. Do you have MongoDB Compass installed on your machine? First thing we should do is ensure your connection to the database is successful.

Note: I'm converting this to a discussion so we can continue troubleshooting there.

**Reply by JamieBradders** [reply]
Hey @JessChowdhury I'm trying Payload for the first time this evening and was getting the same issue when using docker-compose. I can confirm that I've been able to connect to Mongo through TablePlus using the following connection URL:

```
mongodb://127.0.0.1:27017/content-studio
```

However, the logs in Docker are as follows:

```
2023-10-20 22:27:42 content-studio-payload-1  | [nodemon] 2.0.22
2023-10-20 22:27:42 content-studio-payload-1  | [nodemon] to restart at any time, enter `rs`
2023-10-20 22:27:42 content-studio-payload-1  | [nodemon] watching path(s): *.*
2023-10-20 22:27:42 content-studio-payload-1  | [nodemon] watching extensions: ts
2023-10-20 22:27:42 content-studio-payload-1  | [nodemon] starting `ts-node src/server.ts -- -I`
2023-10-20 22:28:14 content-studio-payload-1  | [21:28:13] ERROR (payload): Error: cannot connect to MongoDB. Details: connect ECONNREFUSED 127.0.0.1:27017
2023-10-20 22:28:14 content-studio-payload-1  | [nodemon] app crashed - waiting for file changes before starting...
```

**Reply by JessRynkar** [reply]
Hi @JamieBradders - well that's no good! Let's figure this out...

Few questions to get started...
1. Did you use `yarn` and `yarn dev` to start the server?
2. Do you have an .env file in your root folder? And does it contain a `DATABASE_URI`?
3. What template are you running? If you try another one, do you get the same error?

If you fancy making the repo public, I can take a look directly for you and check if anything looks amiss.

**Reply by mablack** [reply]
@JessChowdhury I'm running into a similar issue with the `ecommerce` template:

```
unhandledRejection Error: mongoose-aggregate-paginate-v2 tried to access mongoose, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.

Required package: mongoose
Required by: mongoose-aggregate-paginate-v2@npm:1.0.6
```

1. Yep, I've run `yarn && yarn dev`
2. I've checked the .env file, and it has the connection string. I've used the same string successfully via MongoDB Compass.
3. I've only checked the `ecommerce` template so far...

### Update

I just ran the following, and it fixed my issue:

```
yarn config set nodeLinker node-modules
```

**Reply by JessRynkar** [reply]
@mablack awesome glad you got it resolved - thanks for sharing your solution here!

**Reply by JamieBradders** [reply]
Sorry @JessChowdhury I didn't see the notification for your reply back in October! I can see that some more workarounds have been posted so I'll try those should I come back to giving Payload another go (I was just tinkering with it before and have since been held up with other jobs). Thank you again!

**Comment by Canowyrms** [comment]
Hi @JessChowdhury , thank you for taking the time to troubleshoot this with me.

I'm not sure when I'll have the opportunity to play around with this again, but if/when I do, I can chime in here again. However, it sounds like @JamieBradders is experiencing the same issue, so if you want to move to troubleshooting with them, please go ahead.

**Reply by JessRynkar** [reply]
Hi @Canowyrms no worries! Feel free to tune back in when you can and let us know if you're still facing issues.

**Comment by henrituan** [comment]
Hi @JessChowdhury,

I'm having the same issue: I got `Error: cannot connect to MongoDB. Details: connect ECONNREFUSED ::1:27017` when running `npm run dev` or ` npm run payload migrate:create -- --file @payloadcms/db-mongodb/versions-v1-v2`

I tried `yarn config set nodeLinker node-modules` too and it didn't work for me.

I am on Window (without WSL). My connection to MongoDB is fine, I can connect with MongoCompass, and with `mongosh` on my terminal. I have this DB connection problem with every version above `2.00`, but with version `1.x.x` it's fine. I don't use any template, just a vanilla payload project from v1.

I am looking forward to migrating to v2, so if you can help troubleshooting I'm grateful. I can provide more information if you needed, unfortunately the log only says `Error: cannot connect to MongoDB. Details: connect ECONNREFUSED ::1:27017`, there is nothing more.

Thanks !

**Reply by henrituan** [reply]
nvm, I found the problem: my mongo connection string should be `mongodb://127.0.0.1:27017/my_db` instead of `mongodb://localhost:27017/my_db`. Don't know why but using localhost in connection string does not work anymore.

**Reply by Canowyrms** [reply]
> Don't know why but using localhost in connection string does not work anymore.

Looks like it was trying to connect on the IPv6 local address (`::1`). Probably Mongo is only binding to IPv4 local address (`127.0.0.1`).

**Reply by 95rade** [reply]
I can't get it to connect, same issue and 'yarn config set nodeLinker node-modules' didn't help resolve it. Mongodb Compass also fails to connect. Mac OS 10.15. Tried 'yarn && yarn dev' as well as 'npm run dev'. No luck. Would really really like to get this resolved.

**Comment by neil-mando** [comment]
### Solution for Docker Compose
Edit `.env` contain line
> `DATABASE_URI=mongodb://127.0.0.1/payload-template-blank`
to
> `DATABASE_URI=mongodb://mongo/payload-template-blank`

Explanation:
- Original config will opening connection to host machine directly to the expose port 27017
- Eventhrough changing from `127.0.0.1` to `0.0.0.0` or `localhost` won't solve it 
- By changing `127.0.0.1` to `mongo` which is docker container host name, it will connect successfully

Test on Payload v2.5.0
Verify by successfully open `http://localhost:3000/admin`

**Reply by fwextensions** [reply]
Props to @neil-mando for delivering the Christmas present that keeps on giving!  Kind of disappointing that Payload's create command is still installing a bad .env example 6 months after the OP's post.

One note: you'll (probably obviously) need to replace `payload-template-blank` with whatever you called your payload project during the scaffolding process.

**Reply by baranelitez** [reply]
i came here in mid-2024 and this is still the solution to this. I think it needs to be integrated into the main branch ASAP.

**Comment by alfredbirk** [comment]
My issue was that I had to install MongoDB. Install it with:
```
brew tap mongodb/brew
brew install mongodb-community
brew services start mongodb-community
```

**Comment by gordeychuk** [comment]
I don't know why, but mongoose doesn't like pnp :) I had the same problem with local mongodb running in docker, but it wasn't working either with 127.0.0.1 or mongodb (or whatever thing I have in /etc/hosts). So my solution was just to switch the nodeLinker:

`yarn config set nodeLinker node-modules`
`yarn install`

Probably it's somehow related to M1 mac.

**Reply by tyteen4a03** [reply]
I don't think *anything* plays well with PnP to be honest.
